src/psyclone/psyad/tl2ad.py

```diff
@@ -88,19 +151,462 @@ def generate_adjoint(tl_psyir):
+
+    # Compute inner product of result of adjoint kernel with original inputs
+    statements += _create_inner_product(inner2,
+                                        zip(inputs, input_copies))
+
+    # Compare the inner products.
+    tol_zero = Literal(INNER_PRODUCT_TOLERANCE, REAL_DOUBLE_TYPE)
+
+    diff = BinaryOperation.create(BinaryOperation.Operator.SUB,
+                                  Reference(inner1), Reference(inner2))
+    abs_diff = UnaryOperation.create(UnaryOperation.Operator.ABS, diff)
+    statements.append(Assignment.create(Reference(diff_sym), abs_diff))
+
+    # If the test fails then the harness will print a message and return early.
+    # TODO #1345 make this code language agnostic.
+    ptree = Fortran2003.Write_Stmt(
+        "write(*,*) 'Test of adjoint of ''{0}'' failed: diff = ', {1}".format(
+            tl_kernel.name, diff_sym.name))
+
+    statements.append(
+        IfBlock.create(BinaryOperation.create(BinaryOperation.Operator.GT,
+                                              Reference(diff_sym),
+                                              tol_zero.copy()),
+                       [CodeBlock([ptree], CodeBlock.Structure.STATEMENT),
+                        Return()]))
+
+    # Otherwise the harness prints a message reporting that all is well.
+    # TODO #1345 make this code language agnostic.
+    ptree = Fortran2003.Write_Stmt(
+        "write(*,*) 'Test of adjoint of ''{0}'' passed: diff = ', {1}".format(
+            tl_kernel.name, diff_sym.name))
+    statements.append(CodeBlock([ptree], CodeBlock.Structure.STATEMENT))
+
+    # Finally, create the driver program from the list of statements.
+    routine = Routine.create(
+        "adj_test", symbol_table, statements, is_program=True)
+
+    logger.debug("Created test-harness program named '{0}'".format(
+        routine.name))
+
+    return routine
+
+
+def _create_inner_product(result, symbol_pairs):
+    '''
+    Creates PSyIR that computes the inner product of each pair of symbols
+    in the supplied list and accumulates it into the supplied `result`
+    variable (which is first zeroed).
+
+    :param result: symbol which will accumulate result.
+    :type result: :py:class:`psyclone.psyir.symbols.DataSymbol`
+    :param symbol_pairs: list of pairs of symbols for which to compute inner \
+                         products.
+    :type symbol_pairs: list of 2-tuples of \
+                        :py:class:`psyclone.psyir.symbols.DataSymbol`
+
+    :returns: PSyIR that performs the inner product and accumulates the \
+        result in the variable represented by the `result` Symbol.
+    :rtype: list of :py:class:`psyclone.psyir.nodes.Assignment`
+
+    :raises TypeError: if any pair of symbols represent different datatypes.
+
+    '''
+    # Zero the variable used to accumulate the result
+    statements = [Assignment.create(Reference(result),
+                                    Literal("0.0", REAL_DOUBLE_TYPE))]
```

Comment:
And here too - please use the proposed defined constant for precision, rather than this explicit value.

Reply:
Done.